README.md
# Manually installing the VanPi system

- Get a clean install of Raspberry Pi OS Lite (Debian 10 Buster)  on a MicroSD-card - ([download here](https://www.raspberrypi.com/software/operating-systems/)) 

-----------------
_**-- IMPORTANT --**_

**_On bullseye the ADS-script (for the shunt etc.) is not working out of the box! Use Buster instead! (The installation script works on both versions, but packages/dependencies needed for future versions of VanPi OS will not be fully installed!)_**
-----------------

 - Using the Raspberry Pi Imager ([download here](https://www.raspberrypi.com/software/)), set the following options
    - Hostname: pekaway.local
    - activate ssh
    - username: pi
    - password: raspberry
    - wifi:
        - Your wifi SSID and passphrase
        - change wifi country if needed
    - change language if needed
    - change keyboardlayout if needed

- Flash the operating system to your sd card


- Put the SD-card into the RPI and power it on (first boot may take a few minutes)
- Wait until it shows up in your network and login via SSH with the credentials you just set
- Once logged in do
```
cd ~/ &&
wget https://git.pekaway.de/Vincent/vanpi/-/raw/main/pi4/vanpi-init.sh &&
chmod +x vanpi-init.sh &&
bash vanpi-init.sh
```

- The script will take about 10-20min to run through, depending on bandwith and hardware (20+ min on RPI3)
- Confirm if inputs are needed (none on bullseye, one on buster -> script works for both)
- Sit back and relax

Once done, the RPI will power up in Access Point Mode, connect to it and proceed from there
(You can just turn off the Access Point because we already configured wifi during the flashing process)

Homebridge needs to be setup, you can do that when accessing RPI-IP:8581 or pekaway.local:8581 - set admin:pekawayfetzt (or whatever you prefer) and continue from there, config should already be there (or you can find it [here](https://git.pekaway.de/Vincent/vanpi/-/blob/main/pi4/config.json)) - not further tested!!
May need a reset through the Node-Red frontend to generate a new random MAC adress before it can be connected to Apple Home

Go to the frontend of the VanPi system > Config > Wifi and click on reset Homebridge. That will generate a new random MAC address, a new Pin code and it'll download the latest homebridge config from our server and deploy it. Remember that you need to be connected to the internet for Apple Home to work (and for downloading the config files, obviously)

## **THROUBLESHOOTING**

In case some symbols in the frontend are not displayed correctly (eg. "°C"), download the flows.json from the pi4 folder ([here](https://git.pekaway.de/Vincent/vanpi/-/blob/main/pi4/flows.json)).
On your RPI open ~/.node-red/flows_pekaway.json and manually replace its content with the content from the file you just downloaded from the server. Then restart Node-Red with "sudo systemctl restart nodered.service".

Alternatively, go into the Node-Red backend and replace non-formatted characters by hand:
 - flow Sensor-Dashboard, on the very top there are 4 text nodes displaying the temp sensors
 - flow Monit, again there are 4 nodes for the temp sensors and 1 gauge node for the CPU-temp
 - flow Config, at "system settings" 4 text nodes for the four temp sensors

 ## On Raspberry Pi Zero W2

npm might not fully install all needed packages (you'll see it when Node-Red reports missing/unknown nodes)
 - login via SSH
 - increase swapsize in /etc/dphys-swapfile to 1024 (sudo nano /etc/dphys-swapfile)
 - reboot
 - sudo systemctl stop nodered.service homebridge.service mosquitto.service nginx.service
 - type "cd ~/.node-red" and then use "npm install --verbose". It'll take like... forever, basically. But it will finish! (Use verbose output to see whats going on)
- sudo systemctl start nodered.service homebridge.service mosquitto.service nginx.service







-----------------------------------------------------------------------------
obsolete
-----------------------------------------------------------------------------
https://git.pekaway.de/Karl/campercontrol





changelog
# Update 1.1.1 (05 May 2023)
- fixed mcp inputs not being displayed correctly
- fixed water level 1 error which kept it from taking other values than 0
- fixed error which prevented "main battdata" to be initialised on boot
- changed sampling frequency of PCA9685 to 200 (dimmer section)
- changed http request nodes in Pekaway Shunt flow to use 127.0.0.1 instead of localhost
- added function to start a second Node-RED instance, using a backup flows file
- added a timezone picker in Frontend (Config > System)
- added function that displays the system time in the info tab & altered format of system time in frontend (Config)

# Update 1.1.0 (31 Mar 2023) >> changes mentioned here also include BETA updates for this version!
- added a precompiled .tft file (VanPI_NSPANEL.tft) for the Sonoff NS Panel to ~/pekaway together with the berry driver (autoexec.be), which is needed for the flashing process
    > quickstart NSPanel:
        > install Tasmota via https://tasmota.github.io/install/ (choose "Tasmota32 Sonoff-NSPanel(englisch)" to be flashed)
        > connect the panel to the wifi accesspoint of your RPI
        > use configs as described here: https://docs.nspanel.pky.eu/prepare_nspanel/
        > set MQTT-Broker to "pekaway.local" and topic to "WifiTouch" (case sensitive)
        > go back to the console in tasmota and type:
            > "Backlog UrlFetch http://pekaway.local/autoexec.be; Restart 1" -> downloads the driver from the local filestorage via nginx
            > "FlashNextion http://pekaway.local/nspaneltft" -> flashes the precompiled .tft file
    > a new flow "Wifi TOUCHPANEL" has been added to Node-RED, which also includes examples for the two hardware buttons on the panel
    > added a function to upload your own .tft file to the raspberry (Config > System > System Update > NS Panel)
    > added functions to install the berry driver and flash the .tft file (Config > System System Update > NS Panel)
- added code to the nginx server block to make the berry driver and the .tft file accessible via http
- added check for i2c devices in dimmer flow to prevent showing error messages if Dimmy is not hardwired
    > messages to PCA9865-Node are limited to 10msg/s (1msg every 0.1s)
- added slider to set max_pwm in % as overvoltage protection (config > switches > dimmer section)
- added dimmytemp sensor data to http endpoint /temp (only when set to active in config)
- reworked the function that generates the default wifi ssid to be a bit more randomized
- reworked the function to reset the wifi to use the same process that generates the default wifi ssid
- reworked the update function 
    > it will now download a script from the Pekaway servers and execute to keep it seperated from Node-RED
    > the script can be found at git.pekaway.de
- fixed error that sometimes displayed ble-bms info in textfield for "currently set battery" (config)
- fixed error with libevdev2:armhf package
______________________________________________________________________________________________________________
# Update 1.1.0 BETA2 (09 Feb 2023)
- added functions to control MaxxAir MaxxFan via IR-Transmitter in combination with ESP device
    > Arduino code still needs to be finalized and published
- disabled all LFBros functions systemwide and purged WiringPi to make UART-Port available
- disabled "gpio readall" function from debug tab
- disabled Dimmer 8 (Dimmy board has only 7 channels available)
- added "dimmytemp1/2" to be set as option for inside temp sensor (2 DS18B20 sensors can be connected to the Dimmy board and will sent their data via MQTT)
- added a PCA9865 node to the dimmer section >> Dimmy can now be connected hardwired without the need for an extra Microcontroller (Wemos)
    > if you want to, you can disable the mqtt outputs for the dimmers and set the slider to send data continously instead of "only on release"
- added function to insert SSID for existing Wifi manually
- fixed errors in ~/ble_py/goksens4s.py (was only showing initial values before)
- fixed function to initialise GokSenso4S on startup
- disabled update via USB function in frontend (was leading to errors in some cases)
- added function to turn off BLE Bat/BMS if not reachable
    > will turn off after 6 failed attempts, then try again after 1 hour (can be turned on again by setting the MAC-adress in the config settings)
    > info: if BLE cannot establish a connection, it will kick out Wifi-clients and block new wifi connections as long as the BLE-requests is running
- prepared flows to use a custom cloud mqtt server (see the flow "MQTT API" at the bottom) 
- added Autoterm 4D to the list of supported heaters (not all status info is delivered via the protocol though)
- added a one-time runtimer to the heater controls >> Start the heater, set the timer, heater will stop when time is over
- changed some settings in the "Chinese Diesel Heater" flow, it now works properly with both types of USB-connectors
- reworked display communication to reduce bandwith
- new beta file for touchdisplay (1.0.3B) is also available

#    new packages/installations
    - node-red-contrib-PCA9865
    - node-red-contrib-countdown-2
______________________________________________________________________________________________________________
# Update 1.1.0 BETA (02 Dec 2022)
- Updated Node-RED core to latest version v3.0.2
 > NOTICE: Node-RED hast to be updated manually ("bash <(curl -sL https://raw.githubusercontent.com/node-red/linux-installers/master/deb/update-nodejs-and-nodered)") when using the update function!
- updated npm packages for Node-RED to latest versions
- updated rpi packages to latest versions ("sudo apt upgrade")
- added link in update checker to read full changelog
- set global variable BatteryDataDelivery to empty ("") on first boot
- added some logic to improve the function that rebuilds the UI
- slightly altered notification message when MainBattData is set to "off"

pi4/updatescript.sh
#!/bin/bash

########################################################
#														
# This script will update your existing VanPi OS 		
# to the latest release from the Pekaway VanPi server!	
# Simply download the script to your VanPi OS and		
# execute it.											
#														
########################################################

# create logfile and make it writable
LOG_FILE=/var/log/pekaway-update.log
sudo touch ${LOG_FILE}
sudo chmod 0666 ${LOG_FILE}

exec 3<&1
coproc mytee { tee ${LOG_FILE} >&3;  }
exec >&${mytee[1]} 2>&1

# define variables
Server='https://git.pekaway.de/Vincent/vanpi/-/raw/main/pi4/'
Version='v1.1.1'		### <--- set new version number VanPi OS
NSPanelVersion='0.0.1'	### <--- set new version number NSPanel
currentVersion=`cat ~/pekaway/version`

# prepare variables to be compared
VersionToCheck='v1.1.1' # Version that has relevant changes in update script
# (if current version number is below that number than this script will execute without the need for confirmation)
VersionSubstring=${VersionToCheck#*v}
currentVersionSubstring=${currentVersion#*v}
IFS='.' read -ra currentVersion_array <<< "$currentVersionSubstring"
IFS='.' read -ra version_array <<< "$VersionSubstring"

currentVersion1=${currentVersion_array[0]}
currentVersion2=${currentVersion_array[1]}
currentVersion3=${currentVersion_array[2]}
newVersion1=${version_array[0]}
newVersion2=${version_array[1]}
newVersion3=${version_array[2]}

echo "This script needs to be run as a sudo user. Press CTRL+C to abort at any time."
echo -e "logfile will be at ${LOG_FILE}"

if [[ $currentVersion1 -lt $newVersion1 ]]; then
  echo "currentVersion ($currentVersion) is older than newVersion ($Version)"
  
elif [[ $currentVersion1 -eq $newVersion1 ]] && [[ $currentVersion2 -lt $newVersion2 ]]; then
  echo "currentVersion ($currentVersion) is older than newVersion ($Version)"
  needUpdate='true'
elif [[ $currentVersion1 -eq $newVersion1 ]] && [[ $currentVersion2 -eq $newVersion2 ]] && [[ $currentVersion3 -lt $newVersion3 ]]; then
  echo "currentVersion ($currentVersion) is older than newVersion ($Version)"
  needUpdate='true'
elif [[ $currentVersion1 -eq $newVersion1 ]] && [[ $currentVersion2 -eq $newVersion2 ]] && [[ $currentVersion3 -eq $newVersion3 ]]; then
 echo "currentVersion ($currentVersion) is the same as newVersion ($Version)"
  needUpdate='false'
else
  echo "currentVersion ($currentVersion) is newer than newVersion ($Version)"
  needUpdate='false'
fi


# get confirmation to continue on manual update
if [[ "$1" == "node-red-auto-update" ]] || [[ "$needUpdate" == "true" ]]; then
	echo -e "proceeding..."
else
	while true; do
		read -r -p "This will update to version ${Version}! Currently you are running version ${currentVersion}. Do you want to continue [y/n]" input
		case $input in
			  [yY][eE][sS]|[yY])
					echo "Ok, proceeding..."
					break;;
			  [nN][oO]|[nN])
					echo "Aborting... bye bye."
					exit
					;;
			  *)
					echo "Invalid input... please type 'y' (yes) or 'n' (no)"
					;;
		esac
	done
fi

# remove packages.txt and package.json if they already exist
cd ~/pekaway
rm -f updatescript.sh
rm -f packages.txt
rm -f package.json
rm -f pip3list.txt
rm -f VanPI_NSPANEL.tft
rm -f autoexec.be

# download new files packages.txt and package.json
echo "downloading new files"
wget ${Server}packages.txt
wget ${Server}package.json
wget ${Server}pip3list.txt
wget ${Server}NSPanel/VanPI_NSPANEL.tft
wget ${Server}NSPanel/autoexec.be

# copy NSPanel .tft file to ~/pekaway/userdata/NSPanel to show up in NR-Dashboard
mkdir -p ~/pekaway/userdata/NSPanel
cp VanPI_NSPANEL.tft ~/pekaway/userdata/NSPanel/VanPI_NSPANEL${NSPanelVersion}.tft

# make a backup of the existing package.json and replace it with the new file
cp ~/.node-red/package.json ~/pekaway/nrbackups/package-backup.json
cp ~/pekaway/package.json ~/.node-red/package.json

# install packages and dependencies
sudo apt update
sudo apt install $(cat ~/pekaway/packages.txt) -y
cd ~/.node-red

# compare older package.json with new one and ask for merging
echo "comparing original package.json with the new one:"

extramodules=$(diff <(jq --sort-keys .dependencies ~/.node-red/package.json) <(jq --sort-keys .dependencies ~/pekaway/nrbackups/package-backup.json) | grep '>')

if [[ -n $extramodules ]]; then
    echo -e "Your original package.json file has the following additonal modules listed:"
	echo -e "$extramodules"

   if [[ "$1" == "node-red-auto-update" ]] || [[ "$needUpdate" == "true" ]]; then
	echo -e "adding additional lines in package.json automatically."
		# cd ~/.node-red
		echo `jq -s '.[0] * .[1]' ~/.node-red/package.json ~/pekaway/nrbackups/package-backup.json` > ~/pekaway/nrbackups/package1.json && jq . ~/pekaway/nrbackups/package1.json > ~/pekaway/nrbackups/pretty.json && rm ~/pekaway/nrbackups/package1.json && mv ~/pekaway/nrbackups/pretty.json ~/pekaway/nrbackups/package1.json
		echo "Missing lines have been added to package.json"
		echo "New package.json:"
		cat ~/.node-red/package.json
   else
		while true; do
			read -r -p "Do you want them to be added to the new package.json? [y/n]" input
			case $input in
				  [yY][eE][sS]|[yY])
						# cd ~/.node-red
						echo `jq -s '.[0] * .[1]' ~/.node-red/package.json ~/pekaway/nrbackups/package-backup.json` > ~/pekaway/nrbackups/package-backup1.json
						jq . ~/pekaway/nrbackups/package-backup1.json > ~/pekaway/nrbackups/pretty.json 
						mv ~/pekaway/nrbackups/package-backup.json ~/pekaway/package-backup.json # keep backup just in case
						rm -f ~/pekaway/nrbackups/package1.json
						mv ~/pekaway/nrbackups/pretty.json ~/.node-red/package.json
						echo "Missing lines have been added to package.json"
						echo "New ~/.node-red/package.json:"
						cat ~/.node-red/package.json
						break
						;;
				  [nN][oO]|[nN])
						echo "No modules added to package.json, proceeding..."
						break
						;;
				  *)
						echo "Invalid input... please type 'y' (yes) or 'n' (no)"
						;;
			esac
		done
	fi
else
    echo "modules are identical, proceeding..."
fi

#install npm modules from package.json
echo "installing npm modules, please stand by..."
npm install
echo "done"

cd ~/pekaway
# Install/update python modules locally (user pi) and globally (root)
echo "installing python modules, please stand by..."
sudo pip3 install -r ~/pekaway/pip3list.txt
sudo pip3 install bottle
pip3 install -r ~/pekaway/pip3list.txt
pip3 install bottle
echo "done"

# remove downloaded files
rm -f ~/pekaway/packages.txt && rm -f ~/pekaway/package.json && rm -f ~/pekaway/pip3list.txt

# backup Node-RED flows
echo "backing up original Node-RED flows"
cd ~/pekaway/nrbackups
cp ~/.node-red/flows_pekaway.json "flows_pekaway_$(date +%d-%m-%Y_%I:%M:%S%p).json"

echo "replacing version number of VanPi OS"
# replace version number
echo ${Version} >| ~/pekaway/version

# set file dimmer_maxpercent to 100
echo "100" > ~/pekaway/dimmer_maxpercent

# set update = true to show up when opening dashboard
echo "1" >| ~/pekaway/update

# download new flows and replace the old file
echo "replacing local flows.json file with new one from the server"
curl ${Server}flows.json > ~/pekaway/pkwUpdate/flows_pekaway.json 
cp ~/pekaway/pkwUpdate/flows_pekaway.json ~/.node-red/flows_pekaway.json
echo "update script finished! You can find the logfile at ${LOG_FILE}."
rm ~/pekaway/pkwUpdate/flows_pekaway.json
sudo systemctl restart nodered.service


pi4/vanpi-init.sh
#!/bin/bash

########################################################
#														
# This script will install VanPi OS				 		
# on a clean and fresh 'Raspberry Pi OS (Buster)'!		
# Simply download the script and execute it.			
#														
########################################################

# create logfile and make it writable
LOG_FILE=/var/log/pekaway-install.log
sudo touch ${LOG_FILE}
sudo chmod 0666 ${LOG_FILE}

exec 3<&1
coproc mytee { tee ${LOG_FILE} >&3;  }
exec >&${mytee[1]} 2>&1

# define startdate
start=`date +%s`
startdate=`date`

# define server address
Server='https://git.pekaway.de/Vincent/vanpi/-/raw/main/pi4/'

# define color variables
Cyan='\033[0;36m'
Red='\033[0;31m'
Yellow='\033[0;33m'
NC='\033[0m' #No Color

# define os-release version
OSRELEASE=`cat /etc/os-release`

echo -e "${Yellow}Logfile will be ${LOG_FILE}${NC}"
# get latest updates
echo -e "${Cyan}updating packages list${NC}"
sudo apt update
echo -e "${Cyan}upgrading packages${NC}"
sudo apt upgrade -y

# enable I2C and 1-Wire
echo -e "${Cyan}Enabling I2C and 1-Wire Bus${NC}"
sudo raspi-config nonint do_i2c 0
echo -e "dtoverlay=w1-gpio\ndtoverlay=uart5\nenable_uart=1" | sudo tee -a /boot/config.txt
sudo debconf-set-selections <<EOF
iptables-persistent iptables-persistent/autosave_v4 boolean true
iptables-persistent iptables-persistent/autosave_v6 boolean true
EOF

# set hostname to pekaway
echo -e "${Cyan}Set hostname to pekaway${NC}"
echo "pekaway" | sudo tee /etc/hostname

# save needed resources to ~/pekaway
echo -e "${Cyan}saving needed resources to ~/pekaway${NC}"
mkdir ~/pekaway
cd pekaway
wget ${Server}home_pi_pekaway_files.zip
unzip ~/pekaway/home_pi_pekaway_files.zip
echo -e "${Cyan}making scripts executable${NC}"
find ~/pekaway/ -type f -iname "*.sh" -exec chmod +x {} \;
chmod 0744 ~/pekaway/availableWifi.txt
chmod 0777 ~/pekaway/pythonsqlite.db
chmod 0755 ~/pekaway/raspi-blinka.py
echo "100" > ~/pekaway/dimmer_maxpercent
echo "yes" > ~/pekaway/firstboot

# get and install needed packages list
echo -e "${Cyan}Saving list with needed packages${NC}"
cd ~/pekaway
wget ${Server}packages_bullseye.txt
wget ${Server}packages_buster.txt
echo -e "${Cyan}Installing needed packages${NC}"

if grep -q "buster" <<< "$OSRELEASE"; then
  echo -e "Installing packages for buster"
  sudo apt-get update
  sudo apt-get upgrade -y
  sudo apt-get install $(cat ~/pekaway/packages_buster.txt) -y
else
  echo -e "Installing packages for bullseye"
  sudo apt install $(cat ~/pekaway/packages_bullseye.txt) -y
  sudo apt install iotop -y
fi
wget ${Server}pip3list.txt
sudo apt install python3-pip -y

# install python modules locally (user pi) and globally (root)
sudo pip3 install -r ~/pekaway/pip3list.txt
sudo pip3 install bottle
pip3 install -r ~/pekaway/pip3list.txt
pip3 install bottle

# download files for NSPanel
cd ~/pekaway
wget ${Server}NSPanel/VanPI_NSPANEL.tft
wget ${Server}NSPanel/autoexec.be
mkdir -p ~/pekaway/userdata/NSPanel
cp ~/pekaway/VanPI_NSPANEL.tft ~/pekaway/userdata/NSPanel/VanPI_NSPANEL.tft
cd ~/

# install git
echo -e "${Cyan}Installing git${NC}"
sudo apt install git make build-essential

# install Node-RED including Node and npm
echo -e "${Cyan}Installing/updating Node-RED, NodeJS and npm${NC}"
cd ~/
bash <(curl -sL https://raw.githubusercontent.com/node-red/linux-installers/master/deb/update-nodejs-and-nodered) --node18 --confirm-install --confirm-pi
sudo systemctl enable nodered.service
echo -e "${Cyan}Starting Node-RED for initial setup...${NC}"
sudo systemctl start nodered.service
echo -e "${Cyan}Sleeping for 20s to let Node-RED do it's thing...${NC}"
sleep 20
echo -e "${Cyan}Stopping Node-RED${NC}"
sudo systemctl stop nodered.service

# install Node-RED modules
echo -e "${Cyan}Installing/updating Node-RED modules...${NC}"
cd ~/.node-red
rm package.json
rm package-lock.json
wget ${Server}package.json
echo -e "${Cyan}Please stand by! This may take a while!${NC}"
echo -e "${Yellow}It may look frozen, but it is not! Please leave it running and wait patiently.${NC}"
echo -e "${Yellow}Go grab a coffee and relax for a while, I'll take care of the rest :)${NC}"
npm install

# install Node-RED Pekaway VanPi flows
echo -e "${Cyan}Installing/updating Node-RED Pekaway VanPi flows...${NC}"
cd ~/pekaway
wget ${Server}flows.json
cp flows.json ~/.node-red/flows_pekaway.json
cd ~/.node-red/node_modules/node-red-dashboard/dist
wget ${Server}icons.zip
mv ~/.node-red/node_modules/node-red-dashboard/dist/icon64x64.png ~/.node-red/node_modules/node-red-dashboard/dist/icon64x64_old.png
mv ~/.node-red/node_modules/node-red-dashboard/dist/icon120x120.png ~/.node-red/node_modules/node-red-dashboard/dist/icon120x120_old.png
mv ~/.node-red/node_modules/node-red-dashboard/dist/icon192x192.png ~/.node-red/node_modules/node-red-dashboard/dist/icon192x192_old.png
unzip icons.zip

# install usbreset.c
echo -e "${Cyan}Installing and compiling usbreset.c...${NC}"
cd ~/
wget ${Server}usbreset.c
sudo gcc usbreset.c -o usbreset
sudo mv usbreset /usr/local/sbin/

# install and configure Access Point
echo -e "${Cyan}Installing and configuring Access Point mode...${NC}"
cd ~/
sudo apt install hostapd dnsmasq -y
sudo systemctl unmask hostapd
sudo systemctl enable hostapd
sudo DEBIAN_FRONTEND=noninteractive apt install -y netfilter-persistent iptables-persistent
sudo cp /etc/dhcpcd.conf /etc/dhcpcd_nonap.conf
echo -e "\n interface wlan0 \n   static ip_address=192.168.4.1/24 \n   nohook wpa_supplicant" | sudo tee -a /etc/dhcpcd.conf
sudo cp /etc/dhcpcd.conf /etc/dhcpcd_ap.conf
echo -e "# https://www.raspberrypi.org/documentation/configuration/wireless/access-point-routed.md \n# Enable IPv4 routing \nnet.ipv4.ip_forward=1" | sudo tee /etc/sysctl.d/routed-ap.conf
sudo iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
sudo netfilter-persistent save
echo -e "# Listening interface\ninterface=wlan0\n# Pool of IP addresses served via DHCP\ndhcp-range=192.168.4.2,192.168.4.20,255.255.255.0,24h\n# Local wireless DNS domain\ndomain=wlan\n# Alias for this router\naddress=/van.pi/192.168.4.1\naddress=/peka.way/192.168.4.1" | sudo tee /etc/dnsmasq.conf
sudo rfkill unblock wlan
echo -e "country_code=DE\ninterface=wlan0\n\nssid=PeKaWayControl\nhw_mode=g\nchannel=6\nmacaddr_acl=0\nauth_algs=1\nwpa=2\nwpa_passphrase=pekawayfetzt\nwpa_key_mgmt=WPA-PSK\nwpa_pairwise=TKIP\nrsn_pairwise=CCMP\n" | sudo tee /etc/hostapd/hostapd.conf

# install and configure Nginx
echo -e "${Cyan}Installing and configuring Nginx${NC}"
sudo apt update && sudo apt install nginx -y
echo -e "server {\n   listen 80;\n   server_name peka.way pekaway.local vanpi.local van.pi;\n   location / {\n      proxy_pass http://127.0.0.1:1880/ui/;\n   }\n}\nserver {\n   listen 80;\n   server_name homebridge.peka.way hb.peka.way homebridge.van.pi hb.van.pi;\n   location / {\n      proxy_pass http://127.0.0.1:8581/;\n   }\n}" | sudo tee /etc/nginx/sites-available/pekaway1
sudo ln -s /etc/nginx/sites-available/pekaway1 /etc/nginx/sites-enabled/
sudo systemctl reload nginx
sudo systemctl enable nginx

# install Mosquitto MQTT Server (probably already installed from packages.txt)
echo -e "${Cyan}Installing Mosquitto MQTT broker${NC}"
sudo apt install mosquitto mosquitto-clients -y
sudo mkdir -p /var/log/mosquitto
sudo touch /var/log/mosquitto/mosquitto.log
sudo chmod 0666 /var/log/mosquitto/mosquitto.log

# implementing new udev rules and restarting udev service
echo -e "${Cyan}Implementing udev rules for serial connections{NC}"
sudo touch /etc/udev/rules.d/98-pekaway-tty.rules
echo -e 'KERNEL=="ttyUSB*", KERNELS=="1-1.1", SYMLINK+="ttyPKW1"\nKERNEL=="ttyUSB*", KERNELS=="1-1.2", SYMLINK+="ttyPKW2"\nKERNEL=="ttyUSB*", KERNELS=="1-1.3", SYMLINK+="ttyPKW3"\nKERNEL=="ttyUSB*", KERNELS=="1-1.4", SYMLINK+="ttyPKW4"' | sudo tee /etc/udev/rules.d/98-pekaway-tty.rules
sudo cp /etc/udev/rules.d/98-pekaway-tty.rules /lib/udev/rules.d/98-pekaway-tty.rules
sudo udevadm control --reload-rules & sudo systemctl restart udev.service

# install Homebridge
echo -e "${Cyan}Installing and configuring Homebridge for Apple Homekit${NC}"
sudo npm install -g --unsafe-perm homebridge homebridge-config-ui-x
sudo hb-service install --user homebridge
echo -e "${Cyan}Installing Mqttthing for Homebridge${NC}"
sudo -E -n npm install -g homebridge-mqttthing@latest
cd ~/pekaway
wget -q ${Server}config.json -O ~/pekaway/config.json
sudo cp -r ~/pekaway/config.json /var/lib/homebridge/config.json

# install Zigbee2MQTT
echo -e "${Cyan}Installing Zigbee2MQTT${NC}"
sudo mkdir /opt/zigbee2mqtt
sudo chown -R ${USER}: /opt/zigbee2mqtt
git clone --depth 1 https://github.com/Koenkk/zigbee2mqtt.git /opt/zigbee2mqtt
cd /opt/zigbee2mqtt && npm ci
echo -e "${Cyan}Downloading config files for Zigbee2MQTT${NC}"
cd ~/pekaway
curl ${Server}configuration.yaml > /opt/zigbee2mqtt/data/configuration.yaml
sudo wget ${Server}zigbee2mqtt.service
sudo mv zigbee2mqtt.service /etc/systemd/system/

# clear files
echo -e "${Cyan}Clearing folders and files...${NC}"
sudo rm ~/pekaway/home_pi_pekaway_files.zip
sudo rm ~/pekaway/packages.txt
sudo rm ~/pekaway/packages_bullseye.txt
sudo rm ~/pekaway/packages_buster.txt
sudo rm ~/pekaway/pip3list.txt
sudo rm ~/pekaway/flows.json

# restart Services
echo -e "${Cyan}Restarting services...${NC}"
sudo systemctl daemon-reload
echo -e "${Cyan}zigbee2mqtt.service is not started/enabled by default!${NC}"
sudo systemctl restart nginx.service homebridge.service mosquitto.service nodered.service bluetooth
sudo chmod 0755 ~/pekaway/ds18b20_py/ds18b20.py
sudo systemctl enable bluetooth
echo -e "${Cyan}Turning off swapfile!${NC}"
sudo swapoff -a
sleep 5
sudo service dphys-swapfile stop
sudo systemctl disable dphys-swapfile

# configure /boot/cmdline.txt
echo -e "${Cyan}Configuring cmdline.txt...${NC}"
sudo sed -i 's/^.*root=PARTUUID/root=PARTUUID/' /boot/cmdline.txt
sed -i 's/flows.json/flows_pekaway.json/g' ~/.node-red/settings.js
sed -i 's/theme: "",/theme: "",\n        header: {\n            title: "Pekaway VAN PI Campercontrol",\n        },/g' ~/.node-red/settings.js
sudo systemctl restart nodered.service

end=`date +%s`
enddate=`date`
runtime=$((end-start))
echo -e "-----------------------------------------"
echo -e "${Cyan}Script started: ${NC}${startdate}"
echo -e "${Cyan}Script ended: ${NC}${enddate}"
echo -e "${Red}Script runtime in Seconds: ${NC}${runtime}"

# reboot Raspberry Pi
echo -e "${Yellow}Installation done, reboot needed!${NC}"
echo -e "${Red}If connection is lost, RPI will reboot into Access Point Mode automatically, \nPlease connect to the access point (VanPiControl_xxx) and proceed from there${NC}"
echo -e "${Red}--> logfile is saved at ${LOG_FILE}${NC}"
echo "yes" > ~/pekaway/firstboot

while true; do
	read -r -p "Do you want to reboot now? [y/n]" input
	 
	case $input in
		  [yY][eE][sS]|[yY])
				sudo reboot
				break;;
		  [nN][oO]|[nN])
				echo "Reboot cancelled, please remember to reboot for the VanPi system to work properly!"
				exit
				;;
		  *)
				echo "Invalid input... please type 'y' (yes) or 'n' (no)"
				;;
	esac
done
